### PR TITLE
add ssh instances on all systems supporting systemd

### PR DESCRIPTION
**Pull request by SimonHoenscheid**
solves #239

### CODE REVIEW

File: metadata.json [hunk 1]
```diff
@@ -15,41 +14,99 @@
     {
       "name": "puppetlabs/concat",
       "version_requirement": ">= 2.2.0 < 8.0.0"
+    },
+    {
+      "name": "puppet/systemd",
+      "version_requirement": ">= 3.6.0 < 4.0.0"
```
Comment (bastelfreak) [inline metadata.json:20]: we released 3.7.0 today:
```suggestion
      "version_requirement": ">= 3.7.0 < 4.0.0"
```

File: manifests/init.pp [hunk 2]
```diff
@@ -98,6 +98,7 @@
 #   Use issue_net header
 #
 class ssh (
+  Hash    $server_instances        = {},
```
Comment (bastelfreak) [inline manifests/init.pp:101]:
```suggestion
  Hash[String[1],Hash[String[1],Variant[String[1],Integer]]] $server_instances = {},
```

File: manifests/params.pp [hunk 3]
```diff
@@ -154,6 +167,8 @@
           $server_package_name = 'openssh'
           $client_package_name = 'openssh'
           $sshd_dir = '/etc/ssh'
+          $sshd_binary = '/usr/sbin/sshd'
```
Comment (bastelfreak) [inline manifests/params.pp:170]: I think we don't need to set the values here when they are already in the `Solaris` block?

File: manifests/server/instances.pp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#   ssh::instances { 'namevar': }
+define ssh::server::instances (
+  String  $ensure                                        = present,
+  Hash    $options                                       = {},
+  String  $service_ensure                                = 'running',
+  Boolean $service_enable                                = true,
+  Boolean $validate_config_file                          = false,
+  Stdlib::Absolutepath $sshd_instance_config_file        = "${ssh::params::sshd_dir}/sshd_config_${title}",
+  Stdlib::Absolutepath $sshd_binary                      = $ssh::params::sshd_binary,
+  Optional[Stdlib::Absolutepath] $sshd_environments_file = $ssh::params::sshd_environments_file,
```
Comment (bastelfreak) [inline manifests/server/instances.pp:15]: Since we include `ssh` we should also rely on the parameters from the main class, not params:
```suggestion
  Stdlib::Absolutepath $sshd_instance_config_file             = "${ssh::sshd_dir}/sshd_config_${title}",
  Stdlib::Absolutepath $sshd_binary                                   = $ssh::sshd_binary,
  Optional[Stdlib::Absolutepath] $sshd_environments_file = $ssh::sshd_environments_file,
```

Comment (bastelfreak) [reply manifests/server/instances.pp:15]: not sure why GitHub renders the indent quite odd :thinking:

File: manifests/server/match_block.pp [hunk 5]
```diff
@@ -7,12 +7,14 @@
   Hash $options  = {},
   String $type   = 'user',
   Integer $order = 50,
+  String $target = $ssh::params::sshd_config,
```
Comment (bastelfreak) [inline manifests/server/match_block.pp:10]:
```suggestion
  Stdlib::Absolutepath $target = $ssh::sshd_config,
```

File: metadata.json [hunk 6]
```diff
@@ -15,41 +14,99 @@
...
     }
   ],
   "operatingsystem_support": [
     {
-      "operatingsystem": "RedHat"
+      "operatingsystem": "RedHat",
+      "operatingsystemrelease": [
+        "7",
```
Comment (bastelfreak) [inline metadata.json:27]: I suggest to add the operating system version numbers in another PR

File: spec/classes/init_spec.rb [hunk 7]
```diff
@@ -1,98 +1,70 @@
...
-      is_expected.to contain_class('ssh::server')
-    end
-    it do
-      is_expected.to contain_concat('/etc/ssh/sshd_config').with_validate_cmd(nil)
-    end
-
-    it do
-      is_expected.to contain_resources('sshkey').with_purge(true)
-    end
+  on_supported_os.each do |os, os_facts|
```
Comment (bastelfreak) [inline spec/classes/init_spec.rb:4]: I suggest to submit the refactoring in another PR

File: spec/classes/init_spec.rb [hunk 8]
```diff
@@ -1,98 +1,70 @@
...
+          }
+        end
 
-      it do
-        is_expected.to contain_class('ssh::client')
+        it { is_expected.not_to contain_resources('sshkey') }
       end
-      it do
-        is_expected.to contain_concat('/etc/ssh/sshd_config').with_validate_cmd('/usr/sbin/sshd -tf %')
+      context 'when on Debian with no other parameters' do
```
Comment (bastelfreak) [inline spec/classes/init_spec.rb:62]: this isn't Debian specific anymore:
```suggestion
      context 'with no other parameters' do
```

**Review by kenyon — CHANGES_REQUESTED** [review]
Just suggesting that new templates be written in EPP instead of ERB.

**Comment by SimonHoenscheid** [comment]
@kenyon bringing EPP to this module is on my list, would do this in a later PR.

**Comment by saz** [comment]
@SimonHoenscheid Tests are failing. Do I need to merge some other PR first?

File: manifests/server/instances.pp [hunk 9]
```diff
@@ -0,0 +1,65 @@
+# @summary A short summary of the purpose of this defined type.
+#
+# A description of what this defined type does
+#
+# @options
+#   Structure see main class
+#
+#   ssh::instances { 'namevar': }
+define ssh::server::instances (
+  String  $ensure                                        = present,
```
Comment (bastelfreak) [inline manifests/server/instances.pp:10]: we should use an enum here or a type from puppetlabs-stdlib

File: manifests/server/instances.pp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  Hash    $options                                       = {},
+  String  $service_ensure                                = 'running',
```
Comment (bastelfreak) [inline manifests/server/instances.pp:12]:
```suggestion
  Stdlib::Ensure::Service  $service_ensure               = 'running',
```

**Comment by saz** [comment]
@SimonHoenscheid Thanks!
